### ISSUE TITLE
request help: proxy-cache plugin, why the cache expiration time cannot be configured dynamically?

### ISSUE THREAD

**Issue by jagerzhang** [apache/apisix#5453]
### Issue description

Nginx 的 proxy-cache 可以动态设置TTL，比如：
```
proxy_cache_valid 200 1m; # 响应码为200的缓存1分钟
proxy_cache_valid 304 1h; # 响应码为304的缓存1小时
...
```
但是APISIX 文档提到：`The cache expiration time cannot be configured dynamically. The expiration time can only be set by the upstream response header Expires or Cache-Control, and the default cache expiration time is 10s if there is no Expires or Cache-Control in the upstream response header`

意思是完全看后端的过期设置来定，否则就按config.yml 默认的cache_ttl，感觉不是那么灵活，是否可以开放下，让我们可以在路由里面自行设置缓存TTL，并可以强制缓存（Nginx的 proxy-cache 还支持proxy_ignore_headers 忽略后端缓存申明...）

### Environment

- apisix version (cmd: `apisix version`): 2.10.0

**Comment by tzssangglass** [comment]
can you show an actual scenario? AFAIK, the cache time is determined by the upstream service and APISIX determines the caching behaviour based on the cache time set by the upstream service, which is equivalent to a protocol. If APISIX breaks this protocol, it can become chaotic.

**Comment by shuaijinchao** [comment]
The setting of the timeout period belongs to the behavior of the application layer. As an intermediate proxy layer, the gateway should be decoupled from the application layer, so I think there is no problem in letting the application layer set the timeout period.

**Comment by jagerzhang** [comment]
@tzssangglass @shuaijinchao 
嗯，从遵循协议上来说你们说的都是有道理的。不过你们应该知道还有一个CDN的场景，是需要支持在网关设置强制缓存的，也就是可以忽略后端应用的缓存申明。而且这一点在Nginx原生的proxy_cache也是支持的，甚至Nginx的还可以支持设置 proxy_cache_ignore_header  来强制忽略后端nocache等不缓存申明。

**Comment by spacewander** [comment]
When I worked in a CDN vendor, I made the proxy_cache_valid time/proxy_cache_ignore_header  configurable via Lua. It requires modifying the Nginx itself.

**Comment by agile6v** [comment]
Hi @jagerzhang Currently You can modify [the basic configuration ](https://github.com/apache/apisix/blob/3fa0c335347a2f980efef94132dfd16c9cce38ea/conf/config-default.yaml#L56)to suit your needs. But it works globally.

**Comment by jagerzhang** [comment]
> Hi @jagerzhang Currently You can modify [the basic configuration ](https://github.com/apache/apisix/blob/3fa0c335347a2f980efef94132dfd16c9cce38ea/conf/config-default.yaml#L56)to suit your needs. But it works globally.

嗯嗯，这个全局设置我知道了~ 不过我还是挺希望APISIX能支持下动态设置过期时间，把这个特性作为插件的一个可配置参数来设置。

这个缓存时长设置在Nginx里面是可以动态设置的，希望可以参考下：http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_valid

**Comment by jagerzhang** [comment]
@spacewander @shuaijinchao @tzssangglass 从APISIX的ngx模板来看，这里也是用的Nginx原生的 `proxy_cache_valid` 指令来缓存的：https://github.com/apache/apisix/blob/fa8a34f72d4de45a42390d17ca27aa9f808deb83/apisix/cli/ngx_tpl.lua#L624

在使用Nginx的时候 `proxy_cache_valid` 是可以支持给指定的响应码定义不同的缓存时间的，也就是说可以配置多条`proxy_cache_valid`指令，比如：
```
proxy_cache_valid 200 30m; # 200响应的请求缓存30分钟
proxy_cache_valid 304 1d; # 304响应的请求缓存1天
....
```
是非常灵活的，其实也非常希望APISIX能继续保留这个特性，可以在插件配置里面灵活的设置不同响应码缓存自定义时长。

**Comment by tzssangglass** [comment]
> 应该是启动的时候就静态生成的

Yes, if you want to do it, @spacewander  way would be better, dynamic.

**Comment by Carl5566** [comment]
這個issue有新的進展嗎？
我也遇到相同的問題，有方法可以解嗎？

**Comment by tzssangglass** [comment]
> 這個issue有新的進展嗎？
> 我也遇到相同的問題，有方法可以解嗎？

now we can use proxy-cache plugin: https://github.com/apache/apisix/blob/master/docs/zh/latest/plugins/proxy-cache.md

**Comment by Carl5566** [comment]
Yes, 我目前也是用proxy-cache plugin, 但是我的需求跟 @jagerzhang 一樣, 希望每條不同的Rooute能有不同的cache ttl, 現在只能所有route都是同一個cache ttl時間

**Comment by Carl5566** [comment]
`{
  "uri": "/*",
  "name": "www1.trustlog.com",
  "methods": [
    "GET",
    "POST",
    "PUT",
    "DELETE",
    "PATCH",
    "HEAD",
    "OPTIONS",
    "CONNECT",
    "TRACE"
  ],
  "host": "www1.trustlog.com",
  "plugins": {
    "proxy-cache": {
      "cache_ttl": 300,
      "disable": false
    },`
我使用了  "cache_ttl": 300, cache的時間依舊是10s

**Comment by tzssangglass** [comment]
> 我使用了 "cache_ttl": 300, cache的時間依舊是10s

```
基于磁盘的缓存需要注意：

不能动态配置缓存的过期时间，只能通过后端服务响应头 Expires 或 Cache-Control 来设置过期时间，如果后端响应头中没有 Expires 或 Cache-Control，那么 APISIX 将默认只缓存 10 秒钟
```

Maybe that's why.

If you think this is a bug, please submit an issue and give the full steps to reproduce it.

**Comment by spacewander** [comment]
So far, `cache_ttl` is not supported by the disk backend. See https://apisix.apache.org/docs/apisix/plugins/proxy-cache#attributes

**Comment by Carl5566** [comment]
請教有什麼方法可以支持不同route 有不同時間的磁盤緩存

**Comment by uran0sH** [comment]
I will work on it. Thanks

**Comment by uran0sH** [comment]
If I'm not mistaken, should I make the proxy-cache plugin support cache_ttl?

**Comment by spacewander** [comment]
Yes

**Comment by uran0sH** [comment]
> Yes

When I read the code of `disk_handler.lua`, I didn't understand how it caches things. It doesn't have the operation like `memory:set`

**Comment by spacewander** [comment]
It uses nginx's proxy-cache

**Comment by uran0sH** [comment]
So do I need to modify the config of Nginx and reload it in `header_filter`? Are there some operations to modify the config of Nginx? I am so confused about it. Could you give me some ideas?

**Comment by spacewander** [comment]
Back from vacation. 

> So do I need to modify the config of Nginx and reload it in `header_filter`? Are there some operations to modify the config of Nginx? I am so confused about it. Could you give me some ideas?

No. We need to modify Nginx and control it via Lua so the config can take effect at the plugin level.

**Comment by uran0sH** [comment]
> Back from vacation.
> 
> > So do I need to modify the config of Nginx and reload it in `header_filter`? Are there some operations to modify the config of Nginx? I am so confused about it. Could you give me some ideas?
> 
> No. We need to modify Nginx and control it via Lua so the config can take effect at the plugin level.

Does it mean that we need to modify the source code of Nginx?

**Comment by Baoyuantop** [comment]
Closing this due to inactivity. If there’s still interest in this feature, please let us know!
